README.md
# 腱鞘炎等级统计器，又名空格计数器

一个用于统计按空格键次数的插件。

## 前言

在一个炎热又无聊的周末，在家玩了两天手机，晚上突然发现右手的大拇指根部酸疼，于是贴了个狗皮膏药以为第二天就能好了。结果第二天还是酸疼酸疼的，于是百度了一下，结果发现好像是腱鞘炎，反思了一下估计是代码敲太多了，而大拇指敲的空格键最多，于是有了开发一个空格统计器的想法。因为敲空格键的声音是 `啪啪啪`，所以本插件名叫 `papapa` 。

## 用法

- 安装之后即可自动运行，会在 vscode 的底部显示今日按空格次数及腱鞘炎段位信息。
- 空格键敲的次数越多，腱鞘炎段位越高。
- 点击底部状态栏可查看你的段位信息。

## 等级

| 次数            | 等级名称 |
| --------------- | -------- |
| 0 - 99999       | 青铜     |
| 100000 - 199999 | 白银     |
| 200000 - 299999 | 黄金     |
| 300000 - 399999 | 铂金     |
| 400000 - 499999 | 钻石     |
| 500000 - 599999 | 大神     |
| 600000 - 699999 | 无敌     |
| >= 700000       | 传说     |

## 源码

https://github.com/LevineHua/vscode-papapa


CHANGELOG.md
# Change Log

All notable changes to the "papapa" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.js
/*
 * @Description: 腱鞘炎计数器，又名空格计数器
 * @Author: Levine
 * @Date: 2022-07-26 18:12:37
 * @LastEditors: Levine
 * @LastEditTime: 2022-07-28 17:12:01
 * @FilePath: /papapa/src/extension.js
 */
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require('vscode');
const { Papapa } = require('./papapa.js');
const info = require('./command/info')

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {

	new Papapa(context)

	console.log('Congratulations, your extension "papapa" is now active!');

	// 注册 papapa.info 命令
	context.subscriptions.push(info(context));
}

// this method is called when your extension is deactivated
function deactivate() {}

module.exports = {
	activate,
	deactivate
}


src/utils/storage.js
/*
 * @Description: 
 * @Author: Levine
 * @Date: 2022-07-27 15:24:53
 * @LastEditors: Levine
 * @LastEditTime: 2022-07-27 17:12:37
 * @FilePath: /papapa/src/utils/storage.js
 */
class Storage {
  constructor(context) {
    this.context = context
  }

  // 获取字符串
  getString(key) {
    if (!key) return 
    console.log('------', key);
    return this.context.globalState.get(key)
  }

  setString(key, value) {
    if (!key) return
    this.context.globalState.update(key, value)
  }

  getObject(key) {
    if (!key) return

    const value = this.getString(key)

    if (!value) return

    return JSON.parse(value, (_key, _value) => {
      const funcPattern = /function[^(]*\(([^)]*)\)\s*\{([\s\S]*)\}/
      const matches = _value ? _value.toString().match(funcPattern) : ''

      if (matches) {
        const args = matches[1].split(',').map((arg) => {
          return arg.trim()
        })
        return new Function(args, matches[2])
      } else {
        return _value
      }
    })
  }

  setObject(
    key,
    value,
    serializable = false,
    rule = {
      type: 'excepts',
      patterns: [],
    }
  ) {
    if (!key) return

    if (typeof value === 'object') {
      value = JSON.stringify(value, (_key, _value) => {
        // 按照规则过滤属性
        if (rule.patterns.indexOf(_key) !== -1) {
          _value = rule.type === 'excepts' ? undefined : _value
        } else {
          _value = rule.type === 'excepts' ? _value : undefined
        }

        if (serializable && typeof _value === 'function') {
          return _value.toString()
        }
        return _value
      })
    } else {
      value = ''
    }

    return this.setString(key, value)
  }
}

module.exports = {
  Storage
}

src/papapa.js
/*
 * @Description: 
 * @Author: Levine
 * @Date: 2022-07-27 15:35:36
 * @LastEditors: Levine
 * @LastEditTime: 2022-07-28 17:39:23
 * @FilePath: /papapa/src/papapa.js
 */
const vscode = require('vscode');
const { Storage } = require('./utils/storage.js')
const { getToDay } = require('./utils/date')
const { LEVELS } = require('./data/level')
const { getLevel } = require('./utils/level')

class Papapa {
  constructor (_context) {
    this.context = _context
    this.storage = new Storage(this.context)
    this.statusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left)
    this.init()
  }

  init() {
    console.log('init=====', this.storage.getObject('papapaNumObj'));
    this.statusBarItem.show()
    this.statusBarItem.command = 'papapa.info'

    this.getToDayPapapaNum()

    var listenner = (event) => {
      if (event) {
        const { contentChanges } = event
        if (contentChanges.length) {
          const text = contentChanges[0].text
          // console.log(text);
          // 判断是否是空格或汉字，如果是汉字则表示输入了空格
          if (text === ' ' || new RegExp("[\\u4E00-\\u9FFF]+", "g").test(text)) {
            // vscode.window.showInformationMessage(papapaNum.toString())
            this.setPapapaNum()
          }
        }
      }
    }
    vscode.workspace.onDidChangeTextDocument(listenner)
  }

  // 设置今天空格的点击数
  setPapapaNum() {
    /**
     * 数据格式：
     * [{
     *  date: '2022-07-27',
     *  num: 1,
     * }]
     */
    // 今天日期
    const date = getToDay()
    // 历史数据
    const papapaNumList = this.storage.getObject('papapaNumObj') || []
    // 判断历史数据中是否存在今天的数据
    const todayDateIndex = papapaNumList.findIndex((val) => val.date === date)
    if (todayDateIndex !== -1) {
      papapaNumList[todayDateIndex].num = papapaNumList[todayDateIndex].num + 1
      this.getToDayPapapaNum(papapaNumList[todayDateIndex].num)
    } else {
      papapaNumList.push({
        date,
        num: 1
      })
      this.getToDayPapapaNum(1)
    }

    this.storage.setObject('papapaNumObj', papapaNumList)
  }

  // 获取今天的空格点击数
  getToDayPapapaNum(todayNum) {
    let num = 0
    if (todayNum) {
      num = todayNum
    } else {
      // // 今天日期
      // const date = getToDay()
      // // 历史数据
      // const papapaNumList = this.storage.getObject('papapaNumObj') || []
      // // 判断历史数据中是否存在今天的数据
      // const todayDateIndex = papapaNumList.findIndex((val) => val.date === date)
      // if (todayDateIndex !== -1) {
      //   num = papapaNumList[todayDateIndex].num
      // }
    }

    // 获取段位
    const { level, today } = getLevel(this.context)
    
    // vscode.window.setStatusBarMessage(`腱鞘炎段位：${level}（今天已敲${num}下空格）`)
    
    this.statusBarItem.text = `腱鞘炎段位：${level}（今天已敲${today || num}下空格）`
  }
}

module.exports = {
  Papapa
}